src/System.Net.WebHeaderCollection/src/System/Net/HeaderInfoTable.cs

```diff
@@ -3,88 +3,146 @@
 
 namespace System.Net
 {
     internal class HeaderInfoTable
     {
-        private static Hashtable HeaderHashTable;
-        private static HeaderInfo UnknownHeaderInfo = new HeaderInfo(string.Empty, false, false, false);
+        private static readonly Func<string, string[]> s_singleParser = value => new[] { value };
+        private static readonly Func<string, string[]> s_multiParser = value => ParseValueHelper(value, isSetCookie: false);
+        private static readonly Func<string, string[]> s_setCookieParser = value => ParseValueHelper(value, isSetCookie: true);
+        private static readonly HeaderInfo s_unknownHeaderInfo = new HeaderInfo(string.Empty, false, false, false, s_singleParser);
+        private static readonly Hashtable s_headerHashTable;
+
+        private static string[] ParseValueHelper(string value, bool isSetCookie)
+        {
+            List<string> tempStringCollection = new List<string>();
+
+            bool inquote = false;
+            int chIndex = 0;
+            char[] vp = new char[value.Length];
+            string singleValue;
+
+            for (int i = 0; i < value.Length; i++)
+            {
+                if (value[i] == '\"')
+                {
+                    inquote = !inquote;
+                }
+                else if ((value[i] == ',') && !inquote)
+                {
+                    singleValue = new string(vp, 0, chIndex);
+                    if (!isSetCookie || !IsDuringExpiresAttributeParsing(singleValue))
+                    {
+                        tempStringCollection.Add(singleValue.Trim());
+                        chIndex = 0;
+                        continue;
+                    }
+                }
+                vp[chIndex++] = value[i];
+            }
+
+            // Now add the last of the header values to the stringtable.
+            if (chIndex != 0) {
+                singleValue = new string(vp, 0, chIndex);
+                tempStringCollection.Add(singleValue.Trim());
+            }
+
+            string[] stringArray = tempStringCollection.ToArray();
+            return stringArray;
+        }
+
+        // This method is to check if we are in the middle of parsing the Expires attribute
+        // for Set-Cookie header. It needs to check two conditions: 1. If current attribute
+        // is Expires. 2. Have we finished parsing it yet. Because the Expires attribute
+        // will contain exactly one comma, no comma means we are still parsing it.
+        private static bool IsDuringExpiresAttributeParsing(string singleValue)
+        {
+            string[] attributeArray = singleValue.Split(';');
+            string lastElement = attributeArray[attributeArray.Length - 1].Trim();
+            bool noComma = lastElement.IndexOf(',') < 0;
+
+            string lastAttribute = lastElement.Split('=')[0].Trim();
+            bool isExpires = string.Equals(lastAttribute, "Expires", StringComparison.OrdinalIgnoreCase);
```

Comment:
There are a bunch of allocations here:
- The string[] and all of the constituent strings from splitting singleValue
- Potentially a string from trimming attributeArray[0]
- The string[] and all of the constituent strings from splitting lastElement

Most of those aren't necessary.

Reply:
I will try to optimize the string allocations in next commit.